tools/pal_inference.py

```diff
@@ -0,0 +1,415 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import argparse
+import copy
+import json
+import logging
+import os
+import signal
+import warnings
+from dataclasses import asdict, dataclass
+from typing import Any, Callable, Dict, List, Optional
+
+import torch
+import torch.nn as nn
+import tqdm
+from datasets import load_dataset
+from transformers import AutoModelForCausalLM, AutoTokenizer
+from transformers.generation.utils import (
+    GenerationConfig,
+    LogitsProcessorList,
+    StoppingCriteriaList,
+)
+from transformers.utils import logging
+
+logger = logging.get_logger(__name__)
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(description="PAL Inference")
+    parser.add_argument("model", type=str)
+    parser.add_argument("out_dir", type=str)
+    parser.add_argument("--dataset", default="gsm8k", type=str)
+    parser.add_argument("--max_length", default=2048, type=int)
+    parser.add_argument("--top_p", default=0.8, type=float)
+    parser.add_argument("--temperature", default=1, type=float)
+    parser.add_argument("--time_out", default=100, type=float)
+    parser.add_argument("--verbose", "-v", action="store_true", help="Print code error information")
+    parser.add_argument("--append", "-a", action="store_true", help="Append output to history results")
+    args = parser.parse_args()
+    return args
+
+
+@torch.inference_mode()
+def generate_interactive(
+    model,
+    tokenizer,
+    prompt,
+    generation_config: Optional[GenerationConfig] = None,
+    logits_processor: Optional[LogitsProcessorList] = None,
+    stopping_criteria: Optional[StoppingCriteriaList] = None,
+    prefix_allowed_tokens_fn: Optional[Callable[[int, torch.Tensor], List[int]]] = None,
+    additional_eos_token_id: Optional[int] = None,
+    **kwargs,
+):
+    inputs = tokenizer([prompt], padding=True, return_tensors="pt")
+    input_length = len(inputs["input_ids"][0])
+    for k, v in inputs.items():
+        inputs[k] = v.cuda()
+    input_ids = inputs["input_ids"]
+    batch_size, input_ids_seq_length = input_ids.shape[0], input_ids.shape[-1]
+    if generation_config is None:
+        generation_config = model.generation_config
+    generation_config = copy.deepcopy(generation_config)
+    model_kwargs = generation_config.update(**kwargs)
+    bos_token_id, eos_token_id = generation_config.bos_token_id, generation_config.eos_token_id
+    if isinstance(eos_token_id, int):
+        eos_token_id = [eos_token_id]
+    if additional_eos_token_id is not None:
+        eos_token_id.append(additional_eos_token_id)
+    has_default_max_length = kwargs.get("max_length") is None and generation_config.max_length is not None
+    if has_default_max_length and generation_config.max_new_tokens is None:
+        warnings.warn(
```

Comment:
logger.warn or warnings.warn, should unified

Reply:
As mentioned in https://github.com/InternLM/InternLM/pull/163#discussion_r1282668580